### PR TITLE
i#3544: RV64: Port test common.getretaddr to RV64

### PR DESCRIPTION
**Pull request by ziyao233**
Port assembly in common/getretaddr.c to riscv64 and implements tailcall_with_retaddr().

Issue: #3544

### CODE REVIEW

File: suite/tests/tools.c [hunk 1]
```diff
@@ -805,8 +805,9 @@ GLOBAL_LABEL(FUNCNAME:)
         mov      x0, x30             /* Replace first argument with return address. */
         br       x9                  /* Tailcall to function pointer. */
 #elif defined(RISCV64)
-        /* TODO i#3544: Port tests to RISC-V64 */
-        ret
+        mv       t0, a0
+        mv       a0, ra
+        jr       t0
```
Comment (ksco) [inline suite/tests/tools.c:810]: Add comments here

**Review by ksco — APPROVED** [review]

File: suite/tests/common/getretaddr.c [hunk 2]
```diff
@@ -113,8 +113,11 @@ GLOBAL_LABEL(FUNCNAME:)
         ldr      x30, [sp], #16
         ret
 # elif defined(RISCV64)
-        /* TODO i#3544: Port tests to RISC-V 64 */
-        ret
+        mv       t0, ra
+        call     next_instr
+    next_instr:
+        mv       a0, ra
```
Comment (abhinav92003) [inline suite/tests/common/getretaddr.c:119]: (No action requested)

I'm a little confused by this routine that's called `get_retaddr` but seems to get the retaddr of an internal call instr that points to the instr at `next_instr`. I don't see how that's useful for the caller.

I guess the precise returned value is not really what the caller wants; maybe the goal is only to verify that this sequence of instrs work under DR (as noted by the "make sure dynamorio can handle this non-call" comment).

**Review by abhinav92003 — APPROVED** [review]

File: suite/tests/tools.c (hunk 1 above)
Comment (ziyao233) [reply suite/tests/tools.c]: Okay

File: suite/tests/common/getretaddr.c (hunk 2 above)
Comment (ziyao233) [reply suite/tests/common/getretaddr.c:119]: > (No action requested)
> 
> I'm a little confused by this routine that's called `get_retaddr` but seems to get the retaddr of an internal call instr that points to the instr at `next_instr`. I don't see how that's useful for the caller.

So do I. The return value isn't really used in the test.

> I guess the precise returned value is not really what the caller wants; maybe the goal is only to verify that this sequence of instrs work under DR (as noted by the "make sure dynamorio can handle this non-call" comment).

**Comment by ziyao233** [comment]
@ksco I've added some comments along with the assembly. This should be ready for merging.

**Pull request merged**